plugins/metrics/epoch_commitments.go

```diff
@@ -0,0 +1,235 @@
+	UTXOsKV, err := storage.WithRealm(kvstore.Realm{storagePrefixUTXOs})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	messagesKV, err := storage.WithRealm(kvstore.Realm{storagePrefixMessages})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	transactionsKV, err := storage.WithRealm(kvstore.Realm{storagePrefixTransactions})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	metrics := &EpochCommitmentsMetrics{
+		epochCommitments:  map[epoch.Index]epoch.ECRecord{},
+		epochVotersWeight: map[epoch.Index]map[identity.ID]float64{},
+		epochUTXOs: objectstorage.NewStructStorage[EpochContent](
+			UTXOsKV,
+			objectstorage.LeakDetectionEnabled(false),
+			objectstorage.StoreOnCreation(true),
+		),
+		epochMessages: objectstorage.NewStructStorage[EpochContent](
+			messagesKV,
+			objectstorage.LeakDetectionEnabled(false),
+			objectstorage.StoreOnCreation(true),
+		),
+		epochTransactions: objectstorage.NewStructStorage[EpochContent](
+			transactionsKV,
+			objectstorage.LeakDetectionEnabled(false),
+			objectstorage.StoreOnCreation(true),
+		),
+	}
+	return metrics, nil
+}
+
+func (m *EpochCommitmentsMetrics) saveCommittedEpoch(ecr *epoch.ECRecord) {
+	m.epochCommitmentsMutex.Lock()
+	defer m.epochCommitmentsMutex.Unlock()
+	m.epochCommitments[ecr.EI()] = *ecr
+}
+
+func (m *EpochCommitmentsMetrics) GetCommittedEpochs() map[epoch.Index]epoch.ECRecord {
+	m.epochCommitmentsMutex.RLock()
+	defer m.epochCommitmentsMutex.RUnlock()
+	duplicate := make(map[epoch.Index]epoch.ECRecord, len(m.epochCommitments))
+	for k, v := range m.epochCommitments {
+		duplicate[k] = v
+	}
+	return duplicate
+}
+
+func (m *EpochCommitmentsMetrics) saveEpochVotersWeight(message *tangle.Message) {
+	branchesOfMessage, err := deps.Tangle.Booker.MessageBranchIDs(message.ID())
+	if err != nil {
+		Plugin.LogError(err)
+	}
+	voter := identity.NewID(message.IssuerPublicKey())
+	var totalWeight float64
+	_ = branchesOfMessage.ForEach(func(txID utxo.TransactionID) error {
+		weight := deps.Tangle.ApprovalWeightManager.WeightOfBranch(txID)
+		totalWeight += weight
+		return nil
+	})
+	m.epochVotersWeightMutex.Lock()
+	defer m.epochVotersWeightMutex.Unlock()
+
+	epochIndex := message.M.EI
+	if _, ok := m.epochVotersWeight[epochIndex]; !ok {
+		m.epochVotersWeight[epochIndex] = map[identity.ID]float64{}
+	}
+	m.epochVotersWeight[epochIndex][voter] += totalWeight
+}
+
+func (m *EpochCommitmentsMetrics) GetEpochVotersWeight() map[epoch.Index]map[identity.ID]float64 {
+	m.epochVotersWeightMutex.RLock()
+	defer m.epochVotersWeightMutex.RUnlock()
+	duplicate := make(map[epoch.Index]map[identity.ID]float64, len(m.epochVotersWeight))
+	for k, v := range m.epochVotersWeight {
+		subDuplicate := make(map[identity.ID]float64, len(v))
+		for subK, subV := range v {
+			subDuplicate[subK] = subV
+		}
+		duplicate[k] = subDuplicate
+	}
+	return duplicate
+}
+
+func (m *EpochCommitmentsMetrics) saveEpochUTXOs(tx *devnetvm.Transaction) {
```

Comment:
I think it would be better to actually obtain the same information the notarization manager has. In fact, you cannot just attach to the TransactionConfirmation, you also have to take into account TransactionInclusionTimeUpdate events and such: and the information could still diverge.
In short: we want to obtain the exact data the notarization manager used to generate the commitment.

Reply:
Done